tests/ci/test_docs_end_to_end/test_runner.py

```
@@ -64,404 +66,326 @@ def _kill_on_timeout() -> None:
+            pass
+
+    reader_thread = threading.Thread(target=_reader, daemon=True)
+    reader_thread.start()
+    reader_thread.join(timeout=timeout)
+    timed_out = reader_thread.is_alive()
+    return lines, timed_out
+
+
+def cleanup_all_containers() -> None:
+    """Stop all running containers and prune stopped ones."""
+    subprocess.run(
```

Ruff (0.16.2)

[error] `subprocess` call with `shell=True` seems safe, but may be changed in the future; consider rewriting without `shell`

(S602)